### PR TITLE
Ft approximate surface

### PR DESCRIPTION
**Pull request by Roxana-P**
# Overview
This pull requests adds a surface approximation routine to the fitting routines. It is implemented according to _The NURBS Book_ by Piegl and Tiller, Section 9.4.3. 
It calculates the least-squares surface approximation to a set of points. This set of points must already have a rectangular structure along the parametric dimensions `num_points_u * num_points_v`, or `num_points_v * num_points_u`.

## Showcase
Surface approximation of a bilinear function f(x,y) = x - y
```
import splinepy
import numpy as np

n_sample = [5, 9]
x, y = np.meshgrid(np.linspace(0,1,n_sample[0]), np.linspace(0, 2, n_sample[1]))
target_points = np.vstack((x.flatten(), y.flatten(), (x - y).flatten())).T
bspline_surface = splinepy.BSpline.approximate_surface(target_points, num_points_u=n_sample[0], num_points_v=n_sample[1], size_u=n_sample[0]-1, size_v=n_sample[1]-5, degree_u=2, degree_v=2)
```

## Checklists
* [x] Documentations are up-to-date.
* [x] Added example(s)
* [x] Added test(s)

### CODE REVIEW

**Comment by j042** [comment]
ah, format changed a bit. you can now use:
```
pip install pre-commit
pre-commit run -a
```

**Comment by j042** [comment]
could you rebase and add this feature in `README.md` under `BSpline fitting` please?

File: splinepy/bspline.py [hunk 1]
```diff
@@ -359,6 +359,96 @@ def interpolate_surface(
...
+          whatever.
+
+        Returns
+        --------
+        fitted: BSpline
+        """
+        if query_points.shape[0] != int(num_points_u * num_points_v):
+            raise ValueError(
+                """The query points do not comply the given
+            number of points."""
```
Comment (j042) [inline splinepy/bspline.py:412]:
```suggestion
                "The query points do not comply the given "
                "number of points."
```

File: splinepy/bspline.py [hunk 2]
```diff
@@ -359,6 +359,96 @@ def interpolate_surface(
...
+            )
+        if (size_u > num_points_u) or (size_v > num_points_v):
+            raise ValueError(
+                """The number of sampling points must be equal
+            or larger than the number of control points in each dimension."""
```
Comment (j042) [inline splinepy/bspline.py:417]:
```suggestion
                "The number of sampling points must be equal "
                "or larger than the number of control points in each dimension."
```

File: tests/test_fitting.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        )
+        interpolated_points = interpolation_spline.sample(
+            resolutions=resolution
+        )
+
+        self.assertTrue(
+            c.np.allclose(
+                f(interpolated_points[:, 0]),
+                interpolated_points[:, 1],
+                atol=0.1,
```
Comment (j042) [inline tests/test_fitting.py:44]: funky

**Review by j042 — COMMENTED** [review]
Awesome, thank you!
What do you think @jzwar?

File: examples/fitting.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    for r in range(0, 6, 1):
+        approximated_surface = splinepy.BSpline.approximate_surface(
+            target_points,
+            num_points_u=n_sample[0],
+            num_points_v=n_sample[1],
+            size_u=n_sample[0] - r,
+            size_v=n_sample[1] - r,
+            degree_u=2,
+            degree_v=2,
+        )
```
Comment (j042) [inline examples/fitting.py:181]: | r | img |
| ------ | ---  |
| 0 | Screenshot 2023-01-05 at 11 43 20 |
| 1 | Screenshot 2023-01-05 at 11 45 31 |
| 2 | Screenshot 2023-01-05 at 11 46 31 |
| ... | ... |
| 5 | Screenshot 2023-01-05 at 11 47 31 |

Comment (j042) [reply examples/fitting.py:181]: Any idea why?

Comment (Roxana-P) [reply examples/fitting.py:154]: Thanks for pointing that out! 
I updated the PR and the examples are also much nicer  now :)

image

File: cpp/splinepy/fitting/fitting.cpp [hunk 5]
```diff
@@ -210,4 +210,103 @@ double ApproximateCurve(const double* points,
...
+
+    ApproximateCurve(pts_u.data(),
+                     num_points_u,
+                     dim,
+                     degree_u,
+                     size_u,
+                     u_k,
+                     knot_vector_u,
+                     coefficient_matrix,
+                     tmp_result);
```
Comment (jzwar) [inline cpp/splinepy/fitting/fitting.cpp:267]: This copy can be avoided if you just pass the correct pointer to the function as they are all aligned in the vector data. Please also check out the next comment

File: cpp/splinepy/fitting/fitting.cpp [hunk 6]
```diff
@@ -210,4 +210,103 @@ double ApproximateCurve(const double* points,
...
+
+  control_points.clear();
+  control_points.assign(size_u * size_v * dim, 0.);
+
+  for (int u = 0; u < size_u; u++) {
+    for (int v = 0; v < num_points_v; v++) {
+      for (int i = 0; i < dim; i++) {
+        pts_v[v * dim + i] = tmp_control_points[(u + (size_u * v)) * dim + i];
+      }
+    }
```
Comment (jzwar) [inline cpp/splinepy/fitting/fitting.cpp:290]: If you replace this move call with a loop and order them in the required order (i.e., change the ordering from u-v-dim to v-u-dim, you can reuse the correct pointer like before and avoid the whole copy operation!

File: cpp/splinepy/fitting/fitting.hpp [hunk 7]
```diff
@@ -36,4 +36,16 @@ double ApproximateCurve(const double* points,
                         std::vector<double>& coefficient_matrix,
                         std::vector<double>& control_points);
 
+void ApproximateSurface(const double* points,
```
Comment (jzwar) [inline cpp/splinepy/fitting/fitting.hpp:39]: Please make this doxygenable (vscode supports it if you just type /** <enter>)

File: cpp/splinepy/py/py_fitting.hpp [hunk 8]
```diff
@@ -202,4 +202,81 @@ py::dict InterpolateSurface(py::array_t<double> points,
...
+                                        num_points_v,  //
+                                        surface_dim,   //
+                                        degree_u,      //
+                                        degree_v,      //
+                                        size_u,        //
+                                        size_v,        //
+                                        centripetal,   //
+                                        knot_vector_u, //
+                                        knot_vector_v, //
+                                        control_points //
```
Comment (jzwar) [inline cpp/splinepy/py/py_fitting.hpp:238]: I know you took this outline from a previous implementation, but once again, the control points could just expect a pointer to the points, as they are all assigned within the function. Here you could just put the pybind data and avoid a copy operation in the end.

**Review by jzwar — COMMENTED** [review]
Hey, nice work. A couple of suggestions. If you deem this to be too much work/unnecessary, we can talk about it and write an issue to push it to the next PR.
Once you checked out all the comments, bump the version push a "Bump version" commit and we can merge this into main

File: cpp/splinepy/fitting/fitting.hpp (hunk 7 above)
Comment (Roxana-P) [reply cpp/splinepy/fitting/fitting.hpp:39]: Sorry, I dont understand this feedback, can you clarify?

**Comment by Roxana-P** [comment]
Thank you for feedback, I included everything except the point about doxygen, which I did not understand. 
Unfortunately, I did not find a way to omit the for loops in l.305-311. 
Moreover, I created two pointers (l. 291, 257), which I do not manually delete. I would be happy about feedback, if that is necessary :)

**Review by jzwar — APPROVED** [review]
Nice, thanks for adding the comments to the header files and changing the implementation. LGTM

**Pull request merged**